code/game/machinery/computer/dropship_weapons.dm

```diff
@@ -196,497 +229,641 @@
+	if(!selected_equipment || !selected_equipment.is_weapon)
+		to_chat(weapon_operator, SPAN_WARNING("No weapon selected."))
+		return FALSE
+	if(dropship.door_override)
+		return FALSE
+	if(!skillcheck(weapon_operator, SKILL_PILOT, DEW.skill_required)) //only pilots can fire dropship weapons.
+		to_chat(weapon_operator, SPAN_WARNING("You don't have the training to fire this weapon!"))
+		return FALSE
+	if(!dropship.in_flyby && DEW.fire_mission_only)
+		to_chat(weapon_operator, SPAN_WARNING("[DEW] requires a fire mission flight type to be fired."))
+		return FALSE
+
+	if(!DEW.ammo_equipped || DEW.ammo_equipped.ammo_count <= 0)
+		to_chat(weapon_operator, SPAN_WARNING("[DEW] has no ammo."))
+		return FALSE
+	if(DEW.last_fired > world.time - DEW.firing_delay)
+		to_chat(weapon_operator, SPAN_WARNING("[DEW] just fired, wait for it to cool down."))
+		return FALSE
+
+	var/datum/cas_iff_group/cas_group = cas_groups[faction]
 
-		var/datum/cas_signal/new_signal
-		for(var/datum/cas_signal/LT as anything in cas_group.cas_signals)
-			if(LT.target_id == targ_id && LT.valid_signal())
-				new_signal = LT
-				break
+	if(!cas_group)
+		return FALSE//broken group. No fighting
 
-		if(!new_signal)
-			to_chat(usr, SPAN_WARNING("Target lost or obstructed."))
-			return
+	for(var/datum/cas_signal/LT in cas_group.cas_signals)
+		if(LT.target_id != targ_id || !LT.valid_signal())
+			continue
+		if(!LT.signal_loc)
+			return FALSE
+		var/turf/TU = get_turf(LT.signal_loc)
+		var/area/targ_area = get_area(LT.signal_loc)
+		var/is_outside = FALSE
+		if(is_ground_level(TU.z))
+			switch(targ_area.ceiling)
+				if(CEILING_NONE)
+					is_outside = TRUE
+				if(CEILING_GLASS)
+					is_outside = TRUE
+		if(!is_outside && !cavebreaker) //cavebreaker doesn't care
+			to_chat(weapon_operator, SPAN_WARNING("INVALID TARGET: target must be visible from high altitude."))
+			return
+		if (protected_by_pylon(TURF_PROTECTION_CAS, TU))
+			to_chat(weapon_operator, SPAN_WARNING("INVALID TARGET: biological-pattern interference with signal."))
+			return FALSE
+		if(!DEW.ammo_equipped.can_fire_at(TU, weapon_operator))
+			return FALSE
+
+		DEW.open_fire(LT.signal_loc)
+		break
+	return TRUE
+
+/obj/structure/machinery/computer/dropship_weapons/proc/ui_create_firemission(mob/weapon_operator, firemission_name, firemission_length)
+	if(!skillcheck(weapon_operator, SKILL_PILOT, SKILL_PILOT_TRAINED)) //only pilots can fire dropship weapons.
+		to_chat(weapon_operator, SPAN_WARNING("A screen with graphics and walls of physics and engineering values open, you immediately force it closed."))
+		return FALSE
+	if(firemission_envelope.max_mission_len <= firemission_envelope.missions.len)
+		to_chat(weapon_operator, SPAN_WARNING("Cannot store more than [firemission_envelope.max_mission_len] Fire Missions."))
+		return FALSE
+	// Check name
+	if(!firemission_name || length(firemission_name) < 5)
+		to_chat(weapon_operator, SPAN_WARNING("Name too short (at least 5 symbols)."))
```

Comment:
This is now a weird requirement especially because there's so little space for the fire mission name to fit in the button. Only 7 characters actually fit in the button, so a shorter name probably ought to be allowed (I don't know if there is any particular reason why it had to be 5 symbols in the first place).
![image](https://github.com/cmss13-devs/cmss13/assets/76988376/be948119-b6b0-4254-a026-46d21e89a6dd)

Maybe should consider allowing word wrap in the button here? Otherwise you could make it fancier by giving it some automatic designation for the button e.g. its "FM-[index]\n[is_multiple_words ? words_abbreviated_up_to_7 : first_7_letters]" so in this case it would be "FM-1\nFH"

Reply:
Also somewhat curious if weird characters like Cyrillic here bust things in any way.